### DISCUSSION TITLE
AVIF with bitdepth 10 or 12 crashes

### DISCUSSION THREAD

**Discussion by LumiCodex** [libvips/libvips#3910] [General]
## Bug report

**Describe the bug**
When I try to use bitdepth at 10 or 12 it crashes vips.exe

It also crashes without throwing exceptions when using with .NET and NETVips

**To Reproduce**
vips copy b.png b.avif[Q=65,bitdepth=10]
![b](https://github.com/libvips/libvips/assets/147385547/e0f56ba5-c68e-4010-a731-322e39f3b4bd)

**Environment**
 - OS: windows 10
 - Vips: vips-dev-w64-all-8.15.2
 
 and 
 - OS: windows 10
 - NetVips 2.4.0
 - NetVips.Native 8.15.1

**Additional context**
I would like to create AVIF 10 bits to use os and browser support for HDR10. Do you know a way to achieve this?

Thank you

**Comment by lovell** [comment]
The prebuilt libvips binaries for Windows provide libaom compiled with support for a bitdepth of 8 only.

If you need support for greater bitdepths, you can build your own binaries via https://github.com/libvips/build-win64-mxe and adjust the `CONFIG_AV1_HIGHBITDEPTH` setting:

https://github.com/libvips/build-win64-mxe/blob/1a4b516c37aed68174d3cc5eda3d1da7e0c507d2/build/aom.mk#L24

**Comment by LumiCodex** [comment]
Thank you it helps a lot, may be it should stile be flagged as a bug as it crashes the app, and doesn't throw exceptions in libraries.

**Reply by lovell** [reply]
I'm not aware of a reliable runtime method to determine if libaom has been compiled with support for high bitdepth images. I guess libvips could have an opt-in compile-time flag to prevent setting `bitdepth` within heifsave.

sharp currently blocks `bitdepth` when using prebuilt binaries - see https://github.com/lovell/sharp/commit/aa1bbcb5c1b871c5d862c66f4cdf1c8dd985c1ed

@kleisauke Would a similar guard in NetVips make sense?

**Reply by LumiCodex** [reply]
I see, do you know why it's restricted to 8 bit when the codec is able to do 12 and is included in the web build?

**Reply by lovell** [reply]
libaom is an exceptionally heavyweight dependency and with high bitdepth support it would add another ~3MB. That _additional_ size is more than is required to fully support _all_ of the JPEG, PNG, GIF and WebP formats combined.

**Reply by LumiCodex** [reply]
I see I will try to do a custom build, with the 12 bit support
Thank a lot!

**Reply by kleisauke** [reply]
> @kleisauke Would a similar guard in NetVips make sense?

That makes sense, but as far as I know there isn't really a way to distinguish pre-built binaries from system-wide binaries without checking if there's a `versions.json` available in the output directory.

I'll try to patch the pre-built binaries to only allow a bitdepth of 8 instead, which would resolve CLI usage as well.

**Reply by LumiCodex** [reply]
if we make a custom build we would stile be able to use the 10 and 12 bits?

**Reply by kleisauke** [reply]
I've just submitted PR https://github.com/libvips/build-win64-mxe/pull/61 for this. This change affects only the pre-built binaries; support for high bitdepths in all other builds remains supported.

**Comment by Auxx** [comment]
It would be great to have a pre-built Windows binary with all features enabled, including HBR for AVIF.

**Reply by kleisauke** [reply]
See: https://github.com/libvips/build-win64-mxe/releases/tag/v8.18.3-rc1. :tada:
